From ca4ebabc1f8a732f051fdce3cca9cf63ae775a5d Mon Sep 17 00:00:00 2001
From: Wei Zhang <zhangwei169@huawei.com>
Date: Fri, 15 Sep 2017 16:26:05 +0800
Subject: [PATCH] key_store:Using euid instead of uid when upgrade wifi blobs

The update transaction goes wrong in "KeyStoreService::upgradeKeyBlob"
function, which try to delete corresponds key blob use "KeyStore.del"
and the target key blob "filename" is generated by uid = "AID_KEYSTORE
1017" which belongs to "android.security.keystore" binder service. '

However, the target key blob is actually owned by uid = "AID_WIFI 1010".
So, we CAN NOT upgrade the wifi password definitely,and connection fail
(could not find that key blob file, because it is not EXIST).

BUG: 65580171

Test: The phone is update to a version with patch level 2017-08.
      1.connect to a wifi hotpoint (Encryption type: "802.1x EAP" ,
        EAP method: :"TLS" , and with "CA certificate"
        and "User certificate" )
      2.update the phone to a new version with patch level 2017-09
      3.connect to the wifi that we have connected in step 1
      4.wifi connect success

Change-Id: I036483b05eb4c5eab2698749069e1118d7f45e48
---
 keystore/key_store_service.cpp | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/keystore/key_store_service.cpp b/keystore/key_store_service.cpp
index f6786b8..a1d8f47 100644
--- a/keystore/key_store_service.cpp
+++ b/keystore/key_store_service.cpp
@@ -1905,6 +1905,13 @@ KeyStoreServiceReturnCode KeyStoreService::upgradeKeyBlob(const String16& name,
             return;
         }
         error = mKeyStore->del(filename.value().string(), ::TYPE_ANY, get_user_id(uid));
+        if (error == ResponseCode::KEY_NOT_FOUND) {
+            uid_t euid = get_keystore_euid(uid);
+            if ((euid != uid) && (euid == AID_WIFI)) {
+                filename = mKeyStore->getBlobFileNameIfExists(name8, euid, ::TYPE_KEYMASTER_10);
+                error = mKeyStore->del(filename.value().string(), ::TYPE_ANY, get_user_id(euid));
+            }
+        }
         if (!error.isOk()) {
             ALOGI("upgradeKeyBlob keystore->del failed %d", (int)error);
             return;
-- 
2.17.1

